src/Microsoft.IdentityModel.Tokens/Base64UrlEncoder.cs

```diff
@@ -207,37 +236,19 @@ internal static unsafe byte[] UnsafeDecode(ReadOnlySpan<char> strSpan)
                     arrayPoolChars = ArrayPool<char>.Shared.Rent(decodedLength);
                 charsSpan = charsSpan.Slice(0, decodedLength);
 
-                source.CopyTo(charsSpan);
-                if (source.Length < charsSpan.Length)
-                {
-                    charsSpan[source.Length] = base64PadCharacter;
-                    if (source.Length + 1 < charsSpan.Length)
-                    {
-                        charsSpan[source.Length + 1] = base64PadCharacter;
-                    }
-                }
-
-                if (needReplace)
-                {
-                    Span<char> remaining = charsSpan;
-                    int pos;
-                    while ((pos = remaining.IndexOfAny(base64UrlCharacter62, base64UrlCharacter63)) >= 0)
-                    {
-                        remaining[pos] = (remaining[pos] == base64UrlCharacter62) ? base64Character62 : base64Character63;
-                        remaining = remaining.Slice(pos + 1);
-                    }
-                }
-
-                source = charsSpan;
+                source = HandlePaddingAndReplace(needReplace, charsSpan, source);
             }
 
             byte[] arrayPoolBytes = null;
             Span<byte> bytesSpan = decodedLength <= StackAllocThreshold ?
                 stackalloc byte[StackAllocThreshold] :
                 arrayPoolBytes = ArrayPool<byte>.Shared.Rent(decodedLength);
 
-            bool converted = Convert.TryFromBase64Chars(source, bytesSpan, out int bytesWritten);
-            Debug.Assert(converted, "Expected TryFromBase64Chars to be successful");
+            int length = Encoding.UTF8.GetBytes(source, bytesSpan);
+            Span<byte> utf8Span = bytesSpan.Slice(0, length);
+            OperationStatus status = System.Buffers.Text.Base64.DecodeFromUtf8InPlace(utf8Span, out int bytesWritten);
```

Comment:
Is the decode operation was not successful (OperationStatus.InvalidData) should we throw an exception?